_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+It wasn't until a few years later that (led by Brooks Davis but with a lot of amazing work from others) we finally had a complete CHERI userspace.
+If you read only one CHERI paper, [the 2019 one describing that system](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201904-asplos-cheriabi.pdf) is probably the one you should read (this one won the ASPLOS Best Paper Award).
+
+# Some things have changed
+
+At the same time, it would be somewhat sad if CHERI had already been perfect in 2015 and it took another decade to get anything into production.
+Quite a few things have changed since then.
+Some are big, some are small refinements.
+
+The most disruptive to the bits of software that I worked on, which I think happened a year or two after this paper, was that the capability and integer register files were merged.
+The original CHERI prototypes used MIPS and used the coprocessor 2 encoding space.
+This made it natural to view the capability unit like a floating-point unit: something with its own register file.
+It turned out that this complicated ABIs, made cores bigger, and didn't really add any benefits.
+Later versions merged the two.
+The MIPS variant maintained both options for a while.
+
+More visibly to the rest of the software stack: Capabilities back then were 256 bits!
+This was great for initial prototypes.
+We originally had a 64-bit base, a 64-bit top, a 64-bit object type, a large space for permissions (31 bits, though most were available for software to use), and some space left over.
+We also had a separate sealed bit, though after a while we realised that this was just an expensive way of encoding the fact that the object type was not zero.
+This was great for a research prototype because there were always spare bits for experiments (though adding the offset required shrinking the object type: we made it 24 bits then, on CHERIoT it's 3 bits and we virtualise it).
+
+Doubling the size of pointers was a hard sell, quadrupling the size would have been impossible.
+It was also painful for large processors to require 256-bit data paths across the load-store unit and register file.
+It took a while to get it published, I can't remember when we first implemented it, but the [CHERI Concentrate paper](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/2019tc-cheri-concentrate.pdf) describes how we shrunk this down to 128 bits, inspired by the [Low-fat pointers work](https://dl.acm.org/doi/10.1145/2508859.2516713).
+CHERIoT had to shrink this further and had other microarchitectural constraints for short pipelines, which is why we support two capability formats for different scales of devices.
+
+And, of course, the MIPS prototypes have been replaced with RISC-V.
+64-bit big-endian MIPS was never well supported and RISC-V reached parity quite quickly and provided a better ecosystem for experimentation.
+
```

Comment:
I wonder if you should also add a section on temporal safety, which was one of the big areas where we “had lots of ideas in the experimental section” but took to fruition from about ISAv7 (2019) onwards, including the path from “track where capabilities go” to “load barrier”, with pointers at the various revocation papers (or at least Cornucopia Reloaded). That also gives another nice comparison point between smaller and larger scale cores?